app/src/main/java/de/westnordost/streetcomplete/data/quest/VisibleQuestsSource.kt

```diff
@@ -91,46 +114,53 @@ class VisibleQuestsSource(
         osmNoteQuestSource.addListener(osmNoteQuestSourceListener)
+        questsHiddenSource.addListener(questsHiddenSourceListener)
         visibleQuestTypeSource.addListener(visibleQuestTypeSourceListener)
         teamModeQuestFilter.addListener(teamModeQuestFilterListener)
         selectedOverlaySource.addListener(selectedOverlayListener)
     }
 
-    fun getAllVisible(bbox: BoundingBox): List<Quest> =
+    fun getAll(bbox: BoundingBox): List<Quest> =
         cache.get(bbox)
 
     /** Retrieve all visible quests in the given bounding box from local database */
-    private fun getAllVisibleFromDatabase(bbox: BoundingBox): List<Quest> {
+    private fun getAllFromDatabase(bbox: BoundingBox): List<Quest> {
+        // we could just get all quests from the quest sources and then filter it with
+        // isVisible(quest) but we can optimize here by querying only quests of types that are
+        // currently visible
         val visibleQuestTypeNames = questTypeRegistry.filter { isVisible(it) }.map { it.name }
         if (visibleQuestTypeNames.isEmpty()) return listOf()
 
-        val osmQuests = osmQuestSource.getAllVisibleInBBox(bbox, visibleQuestTypeNames)
-        val osmNoteQuests = osmNoteQuestSource.getAllVisibleInBBox(bbox)
+        val quests =
+            osmQuestSource.getAllInBBox(bbox, visibleQuestTypeNames) +
+            osmNoteQuestSource.getAllInBBox(bbox)
 
-        return if (teamModeQuestFilter.isEnabled) {
-            osmQuests.filter(::isVisibleInTeamMode) + osmNoteQuests.filter(::isVisibleInTeamMode)
-        } else {
-            osmQuests + osmNoteQuests
-        }
+        return quests.filter { isVisible(it.key) && isVisibleInTeamMode(it) }
     }
 
-    fun get(questKey: QuestKey): Quest? = cache.get(questKey) ?: when (questKey) {
-        is OsmNoteQuestKey -> osmNoteQuestSource.getVisible(questKey.noteId)
-        is OsmQuestKey -> osmQuestSource.getVisible(questKey)
-    }?.takeIf { isVisible(it) }
+    fun get(questKey: QuestKey): Quest? {
+        val quest = cache.get(questKey) ?: when (questKey) {
+            is OsmNoteQuestKey -> osmNoteQuestSource.get(questKey.noteId)
+            is OsmQuestKey -> osmQuestSource.get(questKey)
+        } ?: return null
+        return if (isVisible(quest)) quest else null
+    }
+
+    private fun isVisible(quest: Quest): Boolean =
+        isVisible(quest.key) && isVisibleInTeamMode(quest) && isVisible(quest.type)
 
     private fun isVisible(questType: QuestType): Boolean =
         visibleQuestTypeSource.isVisible(questType) &&
         selectedOverlaySource.selectedOverlay?.let { questType.name !in it.hidesQuestTypes } ?: true
 
-    private fun isVisible(quest: Quest): Boolean =
-        isVisibleInTeamMode(quest) && isVisible(quest.type)
+    private fun isVisible(questKey: QuestKey): Boolean =
+        questsHiddenSource.get(questKey) == null
 
     private fun isVisibleInTeamMode(quest: Quest): Boolean =
         teamModeQuestFilter.isVisible(quest)
```

Comment:
Previously, we checked `teamModeQuestFilter.isEnabled` before calling this function. Is that faster and should it be added back, or does it not matter?

```suggestion
        teamModeQuestFilter.isEnabled && teamModeQuestFilter.isVisible(quest)
```

Reply:
It doesn't matter, the TeamModeQuestFilter checks itself it is enabled in the first line of the `isVisible` function.